dashboard/src/main/home/app-dashboard/validate-apply/services-settings/ServiceList.tsx

```diff
@@ -120,6 +122,14 @@ const ServiceList: React.FC<ServiceListProps> = ({
       {services.length > 0 && (
         <ServicesContainer>
           {services.map((service, idx) => {
+            if (isPredeploy && service.config.type !== "predeploy") {
+              return null;
+            }
+
+            if (!isPredeploy && service.config.type === "predeploy") {
+              return null;
+            }
+
```

Comment:
what is going on here lmao

anyway isn't it cleaner to use a typeguard, maybe defined in services.ts?

typeguard could be:
```
const isNotPredeploy = (service: Service): service is NotPredeployService => {
            return (isPredeploy && service.config.type !== "predeploy") 
( idk what the exact condition in here is)
```

```suggestion
          {services.filter(isNotPredeploy).map((service, idx) => {
         ... core logic
```
though introducing extra types might be annoying

Reply:
Yeah this definitely isn't ideal. One annoying thing with the `react-hook-form` library in this case is that the `useFieldArray` hooks depends on the index of the services in form data to make updates, so if I prefilter it'll end up editing the incorrect services.

I removed the null returns from the component level and just added a flag that checks for predeploy services beforehand.